stripe/src/main/java/com/stripe/android/model/PaymentIntent.java

```diff
@@ -0,0 +1,299 @@
+            String clientSecret,
+            String confirmationMethod,
+            Long created,
+            String currency,
+            String description,
+            Boolean liveMode,
+            Map<String, Object> nextSourceAction,
+            String receiptEmail,
+            ShippingMethod shippingMethod,
+            String source,
+            String status
+    ) {
+        mId = id;
+        mObjectType = objectType;
+        mAllowedSourceTypes = allowedSourceTypes;
+        mAmount = amount;
+        mCanceledAt = canceledAt;
+        mCaptureMethod = captureMethod;
+        mClientSecret = clientSecret;
+        mConfirmationMethod = confirmationMethod;
+        mCreated = created;
+        mCurrency = currency;
+        mDescription = description;
+        mLiveMode = liveMode;
+        mNextSourceAction = nextSourceAction;
+        mReceiptEmail = receiptEmail;
+        mShippingMethod = shippingMethod;
+        mSource = source;
+        mStatus = status;
+    }
+
+    public static String parseIdFromClientSecret(String clientSecret) {
+        return clientSecret.split("_secret")[0];
+    }
+
+    @Nullable
+    public static PaymentIntent fromString(@Nullable String jsonString) {
+        try {
+            return fromJson(new JSONObject(jsonString));
+        } catch (JSONException ignored) {
+            return null;
+        }
+    }
+
+    public static PaymentIntent fromJson(@Nullable JSONObject jsonObject) {
+        if (jsonObject == null || !VALUE_PAYMENT_INTENT.equals(jsonObject.optString(FIELD_OBJECT))) {
+            return null;
+        }
+        String id = optString(jsonObject, FIELD_ID);
+        String objectType = optString(jsonObject, FIELD_OBJECT);
+        JSONArray allowedSourceTypesJSONArray = jsonObject.optJSONArray(FIELD_ALLOWED_SOURCE_TYPES);
+        List<String> allowedSourceTypes = new ArrayList<>();
+        for (int i = 0; i < allowedSourceTypesJSONArray.length(); i++) {
+            try {
+                allowedSourceTypes.add(allowedSourceTypesJSONArray.getString(i));
+            } catch (JSONException ignored) {
+            }
+        }
+        Long amount = optLong(jsonObject, FIELD_AMOUNT);
+        Long canceledAt = optLong(jsonObject, FIELD_CANCELED);
+        String captureMethod = optString(jsonObject, FIELD_CAPTURE_METHOD);
+        String clientSecret = optString(jsonObject, FIELD_CLIENT_SECRET);
+        String confirmationMethod = optString(jsonObject, FIELD_CONFIRMATION_METHOD);
+        Long created = optLong(jsonObject, FIELD_CREATED);
+        String currency = optCurrency(jsonObject, FIELD_CURRENCY);
+        String description = optString(jsonObject, FIELD_DESCRIPTION);
+        Boolean livemode = optBoolean(jsonObject, FIELD_LIVEMODE);
+        String receiptEmail = optString(jsonObject, FIELD_RECEIPT_EMAIL);
+        String status = optString(jsonObject, FIELD_STATUS);
+        String shippingString = optString(jsonObject, FIELD_SHIPPING);
```

Comment:
Fascinating, `optString` serializes the shipping object back to a JSON string?

Reply:
? no, it grabs the String out of the json object?